File: go/a2a_contract_test.go
package haisdk

import (
	"encoding/json"
	"os"
	"reflect"
	"testing"
)

// a2aVerificationContract mirrors the shared canonical fixture structure.
type a2aVerificationContract struct {
	WrappedArtifact           map[string]interface{}   `json:"wrappedArtifact"`
	WrappedArtifactSchema     map[string]interface{}   `json:"wrappedArtifactSchema"`
	VerificationResultSchema  map[string]interface{}   `json:"verificationResultSchema"`
	VerificationResultExample map[string]interface{}   `json:"verificationResultExample"`
	TrustAssessmentSchema     map[string]interface{}   `json:"trustAssessmentSchema"`
	TrustAssessmentExample    map[string]interface{}   `json:"trustAssessmentExample"`
	AgentCardSchema           map[string]interface{}   `json:"agentCardSchema"`
	ChainOfCustodySchema      map[string]interface{}   `json:"chainOfCustodySchema"`
}

func loadA2AContract(t *testing.T) a2aVerificationContract {
	t.Helper()
	data, err := os.ReadFile("../fixtures/a2a_verification_contract.json")
	if err != nil {
		t.Fatalf("read a2a_verification_contract.json: %v", err)
	}
	var contract a2aVerificationContract
	if err := json.Unmarshal(data, &contract); err != nil {
		t.Fatalf("decode a2a_verification_contract.json: %v", err)
	}
	return contract
}

// assertFieldsPresent checks that every expected field from the schema exists in obj.
func assertFieldsPresent(t *testing.T, label string, obj map[string]interface{}, requiredFields map[string]interface{}) {
	t.Helper()
	for field := range requiredFields {
		if field == "_comment" {
			continue
		}
		if _, ok := obj[field]; !ok {
			t.Errorf("%s: missing required field %q", label, field)
		}
	}
}

// assertFieldType checks that a value matches the expected type string.
func assertFieldType(t *testing.T, label, field, expectedType string, value interface{}) {
	t.Helper()
	switch expectedType {
	case "string":
		if _, ok := value.(string); !ok {
			t.Errorf("%s.%s: expected string, got %T", label, field, value)
		}
	case "boolean":
		if _, ok := value.(bool); !ok {
			t.Errorf("%s.%s: expected boolean, got %T", label, field, value)
		}
	case "object":
		if _, ok := value.(map[string]interface{}); !ok {
			t.Errorf("%s.%s: expected object, got %T", label, field, value)
		}
	case "array":
		if _, ok := value.([]interface{}); !ok {
			t.Errorf("%s.%s: expected array, got %T", label, field, value)
		}
	case "number":
		if _, ok := value.(float64); !ok {
			t.Errorf("%s.%s: expected number, got %T", label, field, value)
		}
	}
}

// TestContractWrappedArtifactSerialization verifies that Go's A2AWrappedArtifact
// serializes to JSON with exactly the field names declared in the canonical contract.
func TestContractWrappedArtifactSerialization(t *testing.T) {
	contract := loadA2AContract(t)
	schema := contract.WrappedArtifactSchema
	requiredFields, _ := schema["requiredFields"].(map[string]interface{})

	// Roundtrip the fixture through Go's struct to verify field mapping.
	wrappedJSON, err := json.Marshal(contract.WrappedArtifact)
	if err != nil {
		t.Fatalf("marshal wrapped artifact: %v", err)
	}
	var wrapped A2AWrappedArtifact
	if err := json.Unmarshal(wrappedJSON, &wrapped); err != nil {
		t.Fatalf("unmarshal into A2AWrappedArtifact: %v", err)
	}

	// Re-serialize and check all required fields are present.
	reserialized, err := json.Marshal(wrapped)
	if err != nil {
		t.Fatalf("re-marshal A2AWrappedArtifact: %v", err)
	}
	var reObj map[string]interface{}
	if err := json.Unmarshal(reserialized, &reObj); err != nil {
		t.Fatalf("decode re-serialized: %v", err)
	}

	assertFieldsPresent(t, "A2AWrappedArtifact", reObj, requiredFields)

	// Verify types match schema expectations.
	for field, expectedTypeRaw := range requiredFields {
		if field == "_comment" {
			continue
		}
		expectedType, _ := expectedTypeRaw.(string)
		assertFieldType(t, "A2AWrappedArtifact", field, expectedType, reObj[field])
	}

	// Verify signature sub-fields.
	sigSchema, _ := schema["signatureFields"].(map[string]interface{})
	sigObj, _ := reObj["jacsSignature"].(map[string]interface{})
	if sigObj == nil {
		t.Fatal("A2AWrappedArtifact: jacsSignature is nil after roundtrip")
	}
	assertFieldsPresent(t, "A2AArtifactSignature", sigObj, sigSchema)

	// Specifically check the agentID casing (uppercase ID, not agentId).
	if _, ok := sigObj["agentID"]; !ok {
		t.Error("A2AArtifactSignature: field must be 'agentID' (uppercase ID), not 'agentId'")
	}

	// Check roundtrip values match.
	if wrapped.JacsID != "contract-00000000-0000-4000-8000-000000000001" {
		t.Errorf("jacsId roundtrip: got %q", wrapped.JacsID)
	}
	if wrapped.JacsType != "a2a-task" {
		t.Errorf("jacsType roundtrip: got %q", wrapped.JacsType)
	}
	if wrapped.JacsSignature.AgentID != "contract-agent" {
		t.Errorf("jacsSignature.agentID roundtrip: got %q", wrapped.JacsSignature.AgentID)
	}
}

// TestContractVerificationResultSerialization verifies that Go's
// A2AArtifactVerificationResult serializes with the canonical field names.
func TestContractVerificationResultSerialization(t *testing.T) {
	contract := loadA2AContract(t)
	schema := contract.VerificationResultSchema
	requiredFields, _ := schema["requiredFields"].(map[string]interface{})

	// Roundtrip the example through Go's struct.
	exampleJSON, err := json.Marshal(contract.VerificationResultExample)
	if err != nil {
		t.Fatalf("marshal verification result example: %v", err)
	}
	var result A2AArtifactVerificationResult
	if err := json.Unmarshal(exampleJSON, &result); err != nil {
		t.Fatalf("unmarshal into A2AArtifactVerificationResult: %v", err)
	}

	reserialized, err := json.Marshal(result)
	if err != nil {
		t.Fatalf("re-marshal A2AArtifactVerificationResult: %v", err)
	}
	var reObj map[string]interface{}
	if err := json.Unmarshal(reserialized, &reObj); err != nil {
		t.Fatalf("decode re-serialized: %v", err)
	}

	assertFieldsPresent(t, "A2AArtifactVerificationResult", reObj, requiredFields)

	for field, expectedTypeRaw := range requiredFields {
		if field == "_comment" {
			continue
		}
		expectedType, _ := expectedTypeRaw.(string)
		assertFieldType(t, "A2AArtifactVerificationResult", field, expectedType, reObj[field])
	}

	// Check specific contract values survived roundtrip.
	if result.Valid != false {
		t.Error("valid should be false for placeholder signature")
	}
	if result.SignerID != "contract-agent" {
		t.Errorf("signerId = %q, want %q", result.SignerID, "contract-agent")
	}
	if result.ArtifactType != "a2a-task" {
		t.Errorf("artifactType = %q, want %q", result.ArtifactType, "a2a-task")
	}
}

// TestContractTrustAssessmentSerialization verifies that Go's A2ATrustAssessment
// serializes with the canonical field names.
func TestContractTrustAssessmentSerialization(t *testing.T) {
	contract := loadA2AContract(t)
	schema := contract.TrustAssessmentSchema
	requiredFields, _ := schema["requiredFields"].(map[string]interface{})

	exampleJSON, err := json.Marshal(contract.TrustAssessmentExample)
	if err != nil {
		t.Fatalf("marshal trust assessment example: %v", err)
	}
	var assessment A2ATrustAssessment
	if err := json.Unmarshal(exampleJSON, &assessment); err != nil {
		t.Fatalf("unmarshal into A2ATrustAssessment: %v", err)
	}

	reserialized, err := json.Marshal(assessment)
	if err != nil {
		t.Fatalf("re-marshal A2ATrustAssessment: %v", err)
	}
	var reObj map[string]interface{}
	if err := json.Unmarshal(reserialized, &reObj); err != nil {
		t.Fatalf("decode re-serialized: %v", err)
	}

	assertFieldsPresent(t, "A2ATrustAssessment", reObj, requiredFields)

	for field, expectedTypeRaw := range requiredFields {
		if field == "_comment" {
			continue
		}
		expectedType, _ := expectedTypeRaw.(string)
		assertFieldType(t, "A2ATrustAssessment", field, expectedType, reObj[field])
	}

	// Check specific contract values.
	if assessment.Allowed != true {
		t.Error("allowed should be true for open policy")
	}
	if assessment.TrustLevel != "jacs_verified" {
		t.Errorf("trustLevel = %q, want %q", assessment.TrustLevel, "jacs_verified")
	}
	if assessment.JACSRegistered != true {
		t.Error("jacsRegistered should be true")
	}
	if assessment.InTrustStore != false {
		t.Error("inTrustStore should be false")
	}
	if assessment.Reason != "open policy: all agents accepted" {
		t.Errorf("reason = %q", assessment.Reason)
	}
}

// TestContractAgentCardSerialization verifies that Go's A2AAgentCard
// serializes with the canonical field names from the contract fixture.
func TestContractAgentCardSerialization(t *testing.T) {
	contract := loadA2AContract(t)
	schema := contract.AgentCardSchema
	requiredFields, _ := schema["requiredFields"].(map[string]interface{})

	// Use the existing v04 fixture card as an agent card example.
	cardFixture := loadA2AFixture(t, "agent_card.v04.json")
	cardJSON, err := json.Marshal(cardFixture)
	if err != nil {
		t.Fatalf("marshal card fixture: %v", err)
	}
	var card A2AAgentCard
	if err := json.Unmarshal(cardJSON, &card); err != nil {
		t.Fatalf("unmarshal into A2AAgentCard: %v", err)
	}

	reserialized, err := json.Marshal(card)
	if err != nil {
		t.Fatalf("re-marshal A2AAgentCard: %v", err)
	}
	var reObj map[string]interface{}
	if err := json.Unmarshal(reserialized, &reObj); err != nil {
		t.Fatalf("decode re-serialized: %v", err)
	}

	assertFieldsPresent(t, "A2AAgentCard", reObj, requiredFields)

	for field, expectedTypeRaw := range requiredFields {
		if field == "_comment" {
			continue
		}
		expectedType, _ := expectedTypeRaw.(string)
		assertFieldType(t, "A2AAgentCard", field, expectedType, reObj[field])
	}

	// Check skill sub-fields.
	skillSchema, _ := schema["skillFields"].(map[string]interface{})
	skills, _ := reObj["skills"].([]interface{})
	if len(skills) == 0 {
		t.Fatal("A2AAgentCard: skills array should not be empty")
	}
	firstSkill, _ := skills[0].(map[string]interface{})
	assertFieldsPresent(t, "A2AAgentSkill", firstSkill, skillSchema)

	// Check extension sub-fields.
	extSchema, _ := schema["extensionFields"].(map[string]interface{})
	caps, _ := reObj["capabilities"].(map[string]interface{})
	extensions, _ := caps["extensions"].([]interface{})
	if len(extensions) == 0 {
		t.Fatal("A2AAgentCard: capabilities.extensions should not be empty")
	}
	firstExt, _ := extensions[0].(map[string]interface{})
	// uri is the only truly required field; description and required may be omitted
	if _, ok := firstExt["uri"]; !ok {
		t.Error("A2AAgentExtension: missing required field 'uri'")
	}
	_ = extSchema // used for documentation reference
}

// TestContractChainOfCustodySerialization verifies that Go's A2AChainOfCustody
// serializes with the canonical field names.
func TestContractChainOfCustodySerialization(t *testing.T) {
	contract := loadA2AContract(t)
	schema := contract.ChainOfCustodySchema
	requiredFields, _ := schema["requiredFields"].(map[string]interface{})
	entryFields, _ := schema["entryFields"].(map[string]interface{})

	a2a := mustA2AIntegration(t)

	// Build a minimal chain.
	wrapped := &A2AWrappedArtifact{
		JacsID:          "chain-test-001",
		JacsVersion:     "1.0.0",
		JacsType:        "a2a-task",
		JacsLevel:       "artifact",
		JacsVersionDate: "2026-03-01T00:00:00Z",
		A2AArtifact:     map[string]interface{}{"test": true},
		JacsSignature: &A2AArtifactSignature{
			AgentID:   "chain-agent",
			Date:      "2026-03-01T00:00:00Z",
			Signature: "c2ln",
		},
	}
	chain := a2a.CreateChainOfCustody([]*A2AWrappedArtifact{wrapped})

	chainJSON, err := json.Marshal(chain)
	if err != nil {
		t.Fatalf("marshal chain: %v", err)
	}
	var chainObj map[string]interface{}
	if err := json.Unmarshal(chainJSON, &chainObj); err != nil {
		t.Fatalf("decode chain: %v", err)
	}

	assertFieldsPresent(t, "A2AChainOfCustody", chainObj, requiredFields)

	for field, expectedTypeRaw := range requiredFields {
		if field == "_comment" {
			continue
		}
		expectedType, _ := expectedTypeRaw.(string)
		assertFieldType(t, "A2AChainOfCustody", field, expectedType, chainObj[field])
	}

	// Check chain entry fields.
	entries, _ := chainObj["chainOfCustody"].([]interface{})
	if len(entries) == 0 {
		t.Fatal("A2AChainOfCustody: chainOfCustody array should not be empty")
	}
	firstEntry, _ := entries[0].(map[string]interface{})
	assertFieldsPresent(t, "A2AChainEntry", firstEntry, entryFields)

	for field, expectedTypeRaw := range entryFields {
		if field == "_comment" {
			continue
		}
		expectedType, _ := expectedTypeRaw.(string)
		assertFieldType(t, "A2AChainEntry", field, expectedType, firstEntry[field])
	}
}

// TestContractVerificationResultFieldNames checks that the Go struct's JSON tags
// exactly match the contract fixture field names. This is a compile-time + reflection
// test to catch tag drift.
func TestContractVerificationResultFieldNames(t *testing.T) {
	expected := map[string]string{
		"Valid":            "valid",
		"SignerID":         "signerId",
		"ArtifactType":     "artifactType",
		"Timestamp":        "timestamp",
		"OriginalArtifact": "originalArtifact",
		"Error":            "error",
	}

	typ := reflect.TypeOf(A2AArtifactVerificationResult{})
	for goField, wantTag := range expected {
		field, ok := typ.FieldByName(goField)
		if !ok {
			t.Errorf("A2AArtifactVerificationResult missing Go field %q", goField)
			continue
		}
		tag := field.Tag.Get("json")
		// Strip omitempty etc.
		if idx := len(tag); idx > 0 {
			for i, c := range tag {
				if c == ',' {
					tag = tag[:i]
					break
				}
			}
		}
		if tag != wantTag {
			t.Errorf("A2AArtifactVerificationResult.%s: json tag %q, want %q", goField, tag, wantTag)
		}
	}
}

// TestContractTrustAssessmentFieldNames checks Go struct JSON tags against the contract.
func TestContractTrustAssessmentFieldNames(t *testing.T) {
	expected := map[string]string{
		"Allowed":        "allowed",
		"TrustLevel":     "trustLevel",
		"JACSRegistered": "jacsRegistered",
		"InTrustStore":   "inTrustStore",
		"Reason":         "reason",
	}

	typ := reflect.TypeOf(A2ATrustAssessment{})
	for goField, wantTag := range expected {
		field, ok := typ.FieldByName(goField)
		if !ok {
			t.Errorf("A2ATrustAssessment missing Go field %q", goField)
			continue
		}
		tag := field.Tag.Get("json")
		for i, c := range tag {
			if c == ',' {
				tag = tag[:i]
				break
			}
		}
		if tag != wantTag {
			t.Errorf("A2ATrustAssessment.%s: json tag %q, want %q", goField, tag, wantTag)
		}
	}
}

File: node/tests/a2a-contract.test.ts
/**
 * A2A Verification Contract Tests
 *
 * These tests validate that the Node SDK's A2A types and serialization
 * match the canonical contract fixture at fixtures/a2a_verification_contract.json.
 * They catch schema drift across languages by verifying field names, types,
 * and roundtrip values.
 */
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { resolve } from 'node:path';

// Load the canonical contract fixture.
const CONTRACT_PATH = resolve(__dirname, '../../fixtures/a2a_verification_contract.json');
const contract = JSON.parse(readFileSync(CONTRACT_PATH, 'utf-8'));

type SchemaFields = Record<string, string>;

/** Assert that all required fields from a schema exist in obj. */
function assertFieldsPresent(
  label: string,
  obj: Record<string, unknown>,
  requiredFields: SchemaFields,
) {
  for (const field of Object.keys(requiredFields)) {
    if (field === '_comment') continue;
    expect(obj).toHaveProperty(field);
  }
}

/** Assert that a value matches the expected type string from the schema. */
function assertFieldType(
  label: string,
  field: string,
  expectedType: string,
  value: unknown,
) {
  switch (expectedType) {
    case 'string':
      expect(typeof value).toBe('string');
      break;
    case 'boolean':
      expect(typeof value).toBe('boolean');
      break;
    case 'object':
      expect(value).toBeDefined();
      expect(typeof value).toBe('object');
      expect(Array.isArray(value)).toBe(false);
      break;
    case 'array':
      expect(Array.isArray(value)).toBe(true);
      break;
    case 'number':
      expect(typeof value).toBe('number');
      break;
  }
}

describe('A2A Verification Contract', () => {
  describe('WrappedArtifact schema', () => {
    const wrappedArtifact = contract.wrappedArtifact as Record<string, unknown>;
    const schema = contract.wrappedArtifactSchema as Record<string, unknown>;
    const requiredFields = schema.requiredFields as SchemaFields;
    const signatureFields = schema.signatureFields as SchemaFields;

    it('has all required fields', () => {
      assertFieldsPresent('A2AWrappedArtifact', wrappedArtifact, requiredFields);
    });

    it('required fields have correct types', () => {
      for (const [field, expectedType] of Object.entries(requiredFields)) {
        if (field === '_comment') continue;
        assertFieldType('A2AWrappedArtifact', field, expectedType, wrappedArtifact[field]);
      }
    });

    it('jacsSignature has all required sub-fields', () => {
      const sig = wrappedArtifact.jacsSignature as Record<string, unknown>;
      expect(sig).toBeDefined();
      assertFieldsPresent('A2AArtifactSignature', sig, signatureFields);
    });

    it('uses agentID (uppercase ID) not agentId', () => {
      const sig = wrappedArtifact.jacsSignature as Record<string, unknown>;
      expect(sig).toHaveProperty('agentID');
      // Must NOT have lowercase agentId
      expect(Object.keys(sig)).not.toContain('agentId');
    });

    it('roundtrip values match contract', () => {
      expect(wrappedArtifact.jacsId).toBe('contract-00000000-0000-4000-8000-000000000001');
      expect(wrappedArtifact.jacsType).toBe('a2a-task');
      expect(wrappedArtifact.jacsLevel).toBe('artifact');
      expect(wrappedArtifact.jacsVersion).toBe('1.0.0');
      const sig = wrappedArtifact.jacsSignature as Record<string, unknown>;
      expect(sig.agentID).toBe('contract-agent');
    });
  });

  describe('VerificationResult schema', () => {
    const schema = contract.verificationResultSchema as Record<string, unknown>;
    const requiredFields = schema.requiredFields as SchemaFields;
    const example = contract.verificationResultExample as Record<string, unknown>;

    it('example has all required fields', () => {
      assertFieldsPresent('A2AArtifactVerificationResult', example, requiredFields);
    });

    it('required fields have correct types', () => {
      for (const [field, expectedType] of Object.entries(requiredFields)) {
        if (field === '_comment') continue;
        assertFieldType('A2AArtifactVerificationResult', field, expectedType, example[field]);
      }
    });

    it('example values match contract', () => {
      expect(example.valid).toBe(false);
      expect(example.signerId).toBe('contract-agent');
      expect(example.artifactType).toBe('a2a-task');
      expect(example.timestamp).toBe('2026-03-01T00:00:00Z');
      expect(example.error).toBe('signature verification failed');
    });

    it('uses signerId (camelCase) not signer_id', () => {
      expect(example).toHaveProperty('signerId');
      expect(Object.keys(example)).not.toContain('signer_id');
    });

    it('uses artifactType (camelCase) not artifact_type', () => {
      expect(example).toHaveProperty('artifactType');
      expect(Object.keys(example)).not.toContain('artifact_type');
    });

    it('uses originalArtifact (camelCase) not original_artifact', () => {
      expect(example).toHaveProperty('originalArtifact');
      expect(Object.keys(example)).not.toContain('original_artifact');
    });
  });

  describe('TrustAssessment schema', () => {
    const schema = contract.trustAssessmentSchema as Record<string, unknown>;
    const requiredFields = schema.requiredFields as SchemaFields;
    const example = contract.trustAssessmentExample as Record<string, unknown>;

    it('example has all required fields', () => {
      assertFieldsPresent('A2ATrustAssessment', example, requiredFields);
    });

    it('required fields have correct types', () => {
      for (const [field, expectedType] of Object.entries(requiredFields)) {
        if (field === '_comment') continue;
        assertFieldType('A2ATrustAssessment', field, expectedType, example[field]);
      }
    });

    it('example values match contract', () => {
      expect(example.allowed).toBe(true);
      expect(example.trustLevel).toBe('jacs_verified');
      expect(example.jacsRegistered).toBe(true);
      expect(example.inTrustStore).toBe(false);
      expect(example.reason).toBe('open policy: all agents accepted');
    });

    it('uses trustLevel (camelCase) not trust_level', () => {
      expect(example).toHaveProperty('trustLevel');
      expect(Object.keys(example)).not.toContain('trust_level');
    });

    it('uses jacsRegistered (camelCase) not jacs_registered', () => {
      expect(example).toHaveProperty('jacsRegistered');
      expect(Object.keys(example)).not.toContain('jacs_registered');
    });

    it('uses inTrustStore (camelCase) not in_trust_store', () => {
      expect(example).toHaveProperty('inTrustStore');
      expect(Object.keys(example)).not.toContain('in_trust_store');
    });
  });

  describe('AgentCard schema', () => {
    const schema = contract.agentCardSchema as Record<string, unknown>;
    const requiredFields = schema.requiredFields as SchemaFields;

    // Load the existing v04 card fixture to validate against the schema.
    const cardPath = resolve(__dirname, '../../fixtures/a2a/agent_card.v04.json');
    const card = JSON.parse(readFileSync(cardPath, 'utf-8')) as Record<string, unknown>;

    it('card fixture has all required fields', () => {
      assertFieldsPresent('A2AAgentCard', card, requiredFields);
    });

    it('required fields have correct types', () => {
      for (const [field, expectedType] of Object.entries(requiredFields)) {
        if (field === '_comment') continue;
        assertFieldType('A2AAgentCard', field, expectedType, card[field]);
      }
    });

    it('uses supportedInterfaces (camelCase) not supported_interfaces', () => {
      expect(card).toHaveProperty('supportedInterfaces');
      expect(Object.keys(card)).not.toContain('supported_interfaces');
    });

    it('uses defaultInputModes (camelCase) not default_input_modes', () => {
      expect(card).toHaveProperty('defaultInputModes');
      expect(Object.keys(card)).not.toContain('default_input_modes');
    });

    it('uses defaultOutputModes (camelCase) not default_output_modes', () => {
      expect(card).toHaveProperty('defaultOutputModes');
      expect(Object.keys(card)).not.toContain('default_output_modes');
    });

    it('skills have required sub-fields', () => {
      const skillFields = (schema.skillFields ?? {}) as SchemaFields;
      const skills = card.skills as Record<string, unknown>[];
      expect(skills.length).toBeGreaterThan(0);
      assertFieldsPresent('A2AAgentSkill', skills[0], skillFields);
    });

    it('extensions have uri field', () => {
      const caps = card.capabilities as Record<string, unknown>;
      const extensions = caps.extensions as Record<string, unknown>[];
      expect(extensions.length).toBeGreaterThan(0);
      expect(extensions[0]).toHaveProperty('uri');
    });
  });

  describe('ChainOfCustody schema', () => {
    const schema = contract.chainOfCustodySchema as Record<string, unknown>;
    const requiredFields = schema.requiredFields as SchemaFields;
    const entryFields = schema.entryFields as SchemaFields;

    // Load the golden chain fixture.
    const chainPath = resolve(__dirname, '../../fixtures/a2a/golden_chain_of_custody.json');
    const chainFixture = JSON.parse(readFileSync(chainPath, 'utf-8')) as Record<string, unknown>;
    const expected = chainFixture.expected as Record<string, unknown>;

    it('expected output has schema-defined top-level fields', () => {
      // The golden fixture expected object has entries+totalArtifacts (chain output).
      expect(expected).toHaveProperty('totalArtifacts');
      expect(expected).toHaveProperty('entries');
    });

    it('chain entry fields match schema', () => {
      const entries = expected.entries as Record<string, unknown>[];
      expect(entries.length).toBeGreaterThan(0);
      assertFieldsPresent('A2AChainEntry', entries[0], entryFields);
    });

    it('chain entry fields have correct types', () => {
      const entries = expected.entries as Record<string, unknown>[];
      for (const [field, expectedType] of Object.entries(entryFields)) {
        if (field === '_comment') continue;
        assertFieldType('A2AChainEntry', field, expectedType, entries[0][field]);
      }
    });

    it('uses artifactId (camelCase) not artifact_id', () => {
      const entries = expected.entries as Record<string, unknown>[];
      expect(entries[0]).toHaveProperty('artifactId');
      expect(Object.keys(entries[0])).not.toContain('artifact_id');
    });

    it('uses signaturePresent (camelCase) not signature_present', () => {
      const entries = expected.entries as Record<string, unknown>[];
      expect(entries[0]).toHaveProperty('signaturePresent');
      expect(Object.keys(entries[0])).not.toContain('signature_present');
    });
  });
});
